Feat(spv 1394): Separate admin and user client classes and methods

cain4chain opened the pull request:
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

cain4chain committed:
feat(SPV-1394)- add seperate admin and user clients

feat(admin-api): add admin API client code
feat(user-api): add user API client code
chore(examples): update example code for admin API and user API

cain4chain committed:
feat(SPV-1394)Update admin and user API functions and naming

chore(admin-api user-api): update naming
chore(types)ClientOptions updated for user-api. Update previous ClientOptions to ClientOptionOld naming
chore(examples) Update example files and endpoints back to root folder.

cain4chain committed:
feat(SPV-1394)- Update make requester type definition

fix(user-api)  update make requester types
chore(examples) update tests to xPriv versions

cain4chain committed:
feat(admin-api) update example tests

chore(webhook-example)  Adjust webhook example test for get webhooks

cain4chain committed:
feat(SPV-1394) : Update all test files and webhooks to new APIs

chore(jest): update all test files
chore(webhooks): update webhooks to new admin API functions
chore(README.md): update readme documentation regarding new admin and user endpoints

cain4chain committed:
Merge remote-tracking branch 'origin/main' into feat(SPV-1394)-Separate-admin-and-user-client-classes-and-methods

cain4chain committed:
feat(SPV-1394): resolve conflicts

feat(conflicts): update file conflicts

dorzepowski commented on src/user_api.ts:
```diff
@@ -0,0 +1,553 @@
+     * @param {ClientOptions} options - Configuration options including xPub, xPriv, or accessKey
+     * @param {LoggerConfig} loggerConfig - Logger configuration (optional)
+     */
+    constructor(serverUrl: string, options: ClientOptions, loggerConfig: LoggerConfig = defaultLogger) {
```
maybe it's good idea to move logger also into ClientOptions. WDYT?

dorzepowski commented on src/user_api.ts:
```diff
@@ -0,0 +1,553 @@
+      return await this.http.request(`${basePath}${queryString}`, 'GET');
+    }
+  }
+  
```
by some reason, it is still showing, that there is not empty line at the end,
maybe it's wrong character used to create new line, 
maybe try to remove it and commit again 🤷

cain4chain replied:
I will check this again thanks

dorzepowski commented on src/admin_api.ts:
```diff
@@ -0,0 +1,374 @@
+     * @param {string} adminKey - Admin xPriv key
+     * @param {LoggerConfig} loggerConfig - Logger configuration (optional)
+     */
+    constructor(serverUrl: string, adminKey: string, loggerConfig: LoggerConfig = defaultLogger) {
```
Maybe it should be similar to consturctor for user.
For example something like:
```suggestion
    constructor(serverUrl: string, options AdminClientOptions) {
```

where AdminClientOptions would be
```typescript
interface AdmincClientOptions extends XprivWithSigning {
   logger: LoggerConfig
}
```

_(maybe  `{ logger: LoggerConfig }` should also have it's type.)_

cain4chain replied:
Yes makes sense, keeps it consistent  with the userAPI, and allows extension of options later if required such as admin access key etc.

dorzepowski commented on examples/xpub-from-xpriv.ts:
```diff
@@ -1,6 +1,6 @@
 // This is an example xPriv key - replace it with your own
-const xPriv = 'xprv9s21ZrQH143K4VneY3UWCF1o5Kk2tmgGrGtMtsrThCTsHsszEZ6H1iP37ZTwuUBvMwudG68SRkcfTjeu8h3rkayfyqkjKAStFBkuNsBnAkS';
-
+const xPriv =
+  'xprv9s21ZrQH143K3CbJXirfrtpLvhT3Vgusdo8coBritQ3rcS7Jy7sxWhatuxG5h2y1Cqj8FKmPp69536gmjYRpfga2MJdsGyBsnB12E19CESK';
```
hmmm... I don't think that we nned to change this

cain4chain replied:
I will change this back to the original, thanks

dorzepowski commented on src/client.ts:
We should remove this whole file, but you can do that in a separate PR.

dzolt-4chain replied:
If you decide to remove it in this PR I would also suggest to bump client version to beta.32 in `package.json`

cain4chain replied:
Yes this file can be removed now as part of the new version. Thanks

dorzepowski commented on src/admin_api.ts:
```diff
@@ -0,0 +1,374 @@
+     *
+     * @returns {Promise<Webhook[]>} List of webhook subscriptions
+     */
+    async getWebhooks(): Promise<Webhook[]> {
```
to meet the convention in method names here, it shouldn't have `get` prefix in name
 ```suggestion
    async webhooks(): Promise<Webhook[]> {
```

cain4chain replied:
No problem, will update. Thanks

dorzepowski commented:
One small tip.
To make it simpler to release new version of this library, I would suggest to change `version` number in package.json to the next version number in this PR.

The release of new version after merge is all about pushing a git tag with new version number, then gh actions will try to release that version for you, but if the git tag won't match version from package.json, then the release will fail.
Unfortunately we don't have the automation for changing version in package.json, so you need to do that manually. And if you won't do that in this PR, you would need to do that in a separate PR.

chris-4chain commented on README.md:
```diff
@@ -50,22 +50,25 @@ yarn add @bsv/spv-wallet-js-client
   xPriv: 'xpriv.....',
 });
 
+// Create a new instance of the SPV Wallet admin client
+const adminClient = new SPVWalletAdminAPI(spvWalletServerUrl, 'adminKey.....');
+
```
I would put the `adminClient` creation into another snippet. 
Because now, it seems like it breaks the example how to use userClient.
Now we have:
```js
1. userClient creation

2. adminClient creation (for no reason)

3. Get back to userClient
const userInfo = await userClient.xPub();
console.log('Current balance:', userInfo.currentBalance);

// No usage of adminClient at all
```

I would suggest to split it into `two` code examples with proper description:
```js
1. userClient creation

2. example how to use
const userInfo = await userClient.xPub();
console.log('Current balance:', userInfo.currentBalance);
```

```js
1. adminClient creation

2. example how to use adminClient
// for instance, get stats 
```

chris-4chain commented on README.md:
```diff
@@ -76,52 +79,42 @@ console.log('Current balance:', userInfo.currentBalance);
-  adminKey: 'adminkey.....',
-});
+
+const adminClient = new SPVWalletAdminAPI(spvWalletServerUrl, 'adminkey.....');
```
Maybe, additional lines with "how to use this adminClient" would be good :)

chris-4chain commented on examples/admin-remove-user.ts:
```diff
@@ -9,8 +9,6 @@ if (!exampleAdminKey) {
+const adminClient = new SPVWalletAdminAPI(server, exampleAdminKey);
 
-await adminClient.AdminDeletePaymail('d43ed481ba08aae1db02d880ebefe962f9796168387bb293a95024cb02b953ef');
+await adminClient.deletePaymail(examplePaymail, examplePaymail);
```
I looks weird, why you changed one-argument-method (with some ID (I guess)) into two-arguments-method with `examplePaymail, examplePaymail`. 

Maybe, you made changes on "old" codebase and while updating the branch you didn't resolve conflicts properly 🤷

chris-4chain replied:
Ok, I got this - here is a merged PR with related changes. https://github.com/bitcoin-sv/spv-wallet-js-client/pull/1137

If you have questions about it, feel free to ask @dzolt-4chain 🙈

cain4chain replied:
Ahhh ok, will check this thanks

chris-4chain commented on examples/custom-logger.ts:
```diff
@@ -17,7 +16,7 @@ const clientWithCustomLogLevel = new SpvWalletClient(
 
 //this will log that the request is being made
-await clientWithCustomLogLevel.GetTransactions({}, {}, {});
+await clientWithCustomLogLevel.transactions({}, {}, {});
```
Could you check if these empty objects are necessary?
- Does this work like this? 
```
clientWithCustomLogLevel.transactions();
```

dzolt-4chain replied:
It won't work like that... now. 

In order for that to work we need to change the method implementations, which is not hard btw.

The change would require to do something like this: 

```js
async transactions(
    conditions: TransactionFilter = {},
    metadata: Metadata = {},
    params: QueryPageParams = {},
  ): Promise<PageModel<AdminTx>> {
    const basePath = 'admin/transactions';
    const queryString = buildQueryPath({
      filter: conditions,
      metadata,
      page: params,
    });
    return await this.http.adminRequest(`${basePath}${queryString}`, 'GET');
  }
```

In JavaScript, you can skip arguments in a function by providing default values, such as = {}. However, when using default values, you can only omit arguments from right to left. This means you cannot selectively skip optional arguments in the middle of the parameter list without also providing values for the arguments that come after them.
```js
function example(a, b = {}, c = {}) {
    console.log(a, b, c);
}

// You can skip `c` by only providing `a` and `b`:
example(1, { key: 'value' }); // Output: 1, { key: 'value' }, {}

// But you cannot skip `b` while providing `c` without explicitly passing `undefined` for `b`:
example(1, undefined, { key: 'value' }); // Output: 1, {}, { key: 'value' }
```

wregulski commented on examples/generate-totp.ts:
```diff
@@ -0,0 +1,32 @@
+import { SPVWalletUserAPI } from '../dist/typescript-npm-package.cjs.js';
```
This file has been replaced by `manage-contacts.ts` so we can remove it.

wregulski commented on examples/package.json:
```diff
@@ -7,24 +7,25 @@
-    "sync-merkleroots": "yarn run:example sync-merkleroots.ts",
-    "update-user-xpub-metadata": "yarn run:example update-user-xpub-metadata.ts",
+    "admin-get-handled-paymail-domains": "yarn run:example admin-get-handled-paymail-domains.ts",
+    "user-generate-totp": "yarn run:example generate-totp.ts",
```
This line is unnecessary, we should use:
`"manage-contacts": "yarn run:example manage-contacts.ts",` instead.

wregulski commented on src/admin_api.ts:
```diff
@@ -0,0 +1,374 @@
+      return await this.http.adminRequest('configs/shared', 'GET');
+    }
+  }
+  
```
minor, please add new line at the end of the file 🙈

cain4chain replied:
thanks, sorting :)

dzolt-4chain commented on examples/admin-remove-user.ts:
```diff
@@ -9,8 +9,6 @@ if (!exampleAdminKey) {
+const adminClient = new SPVWalletAdminAPI(server, exampleAdminKey);
 
-await adminClient.AdminDeletePaymail('d43ed481ba08aae1db02d880ebefe962f9796168387bb293a95024cb02b953ef');
+await adminClient.deletePaymail(examplePaymail, examplePaymail);
```
Delete paymail now uses only id, thats why that string was in the examples.

cain4chain replied:
ok thanks, will update

dzolt-4chain commented on src/admin_api.ts:
```diff
@@ -0,0 +1,374 @@
+    /**
+     * Delete a paymail
+     */
+    async deletePaymail(id: string, address: string): Promise<void> {
```
Delete paymail only accepts id in it's path param, we no longer require address in body. Please delete it. 
and please add in the function description `@param {string} id`

cain4chain replied:
Thanks, will update this.

dzolt-4chain commented on src/index.ts:
```diff
@@ -6,3 +6,5 @@ export * from './authentication';
 export * from './types';
 export * from './errors';
 export * from './notifications';
+export * from './admin_api';
```
Since all the files were written in kebab-case maybe it would be better to actually rename admin_api and user_api to kebab-case? WDYT?

cain4chain replied:
I used naming comparisons to go client but this is a good point I will update this. Thanks

dzolt-4chain commented on src/regression_tests/utils.ts:
```diff
@@ -76,48 +77,51 @@ export const createUser = async (paymail: string, paymailDomain: string, instanc
-  const adminClient = new SpvWalletClient(instanceURL, { adminKey: adminXPriv, xPriv: adminXPriv });
-  await adminClient.AdminDeletePaymail(paymailId);
+export const removeRegisteredPaymail = async (
+  paymail: string,
```
After you remove paymail from deletePaymail please also remove it from here. Now regression tests won't pass because in regresiontests.test file we are only providing paymailId without the paymail address itself, which is the correct behaviour

cain4chain replied:
Understood, will update. Thanks

dzolt-4chain commented on src/regression_tests/utils.ts:
```diff
@@ -76,48 +77,51 @@ export const createUser = async (paymail: string, paymailDomain: string, instanc
+  adminXPriv: string,
+) => {
+  const adminClient = new SPVWalletAdminAPI(instanceURL, adminXPriv);
+  await adminClient.deletePaymail(paymailId, paymail);
```
```suggestion
  await adminClient.deletePaymail(paymailId);
```

cain4chain committed:
feat(SPV-1394) Update Admin API options

feat(admin-api): add AdminClientOptions for admin API
docs(readme): update docs with admin and client separation description
test(examples): update example test files with new admin api changes
chore(package.json): update beta versioning to 3.2

chris-4chain committed:
Merge branch 'main' into feat(SPV-1394)-Separate-admin-and-user-client-classes-and-methods

cain4chain committed:
feat(SPV-1394): update example test naming

chore(examples): update test command to match example test file naming
